### PR TITLE
Move to weakref cache.

### PR DESCRIPTION
**Pull request by hameerabbasi**
I'm getting no performance benefits from this cache, and I have verified that the cache is hit most of the time.

### CODE REVIEW

**Comment by hameerabbasi** [comment]
Note that this PR is for discussion, not for merging.

File: uarray/_backend.py [hunk 1]
```diff
@@ -191,7 +186,17 @@ def set_backend(backend, coerce=False, only=False):
     skip_backend: A context manager that allows skipping of backends.
     set_global_backend: Set a single, global backend for a domain.
     """
-    return _SetBackendContext(backend, coerce, only)
+    try:
+        return _set_backend_cache[weakref.ref(backend), coerce, only]
```
Comment (peterbell10) [inline uarray/_backend.py]: We can avoid the need to construct the `weakref` object by implementing in C++ where we can key by `PyObject*` without holding a reference count.

File: uarray/_backend.py [hunk 2]
```diff
@@ -191,7 +186,17 @@ def set_backend(backend, coerce=False, only=False):
...
+    except KeyError:
+
+        def del_attr(x):
+            del _set_backend_cache[x, coerce, only]
+
+        ref = weakref.ref(backend, del_attr)
+        ctx = _SetBackendContext(backend, coerce, only)
+        _set_backend_cache[ref, coerce, only] = ctx
```
Comment (peterbell10) [inline uarray/_backend.py]: Note that the context object itself holds an owning reference to the backend; so this still leaks references.  I think this is the hardest problem to solve.

Comment (hameerabbasi) [reply uarray/_backend.py]: If the context object exists, we need the reference to the backend anyway.

File: uarray/_backend.py (hunk 1 above)
Comment (hameerabbasi) [reply uarray/_backend.py]: How do we ensure we delete when the object goes out of scope?

File: uarray/_backend.py (hunk 2 above)
Comment (hameerabbasi) [reply uarray/_backend.py]: Note also that the problem we're trying to solve is to GC the context, not the backend.

File: uarray/_backend.py (hunk 1 above)
Comment (peterbell10) [reply uarray/_backend.py]: Since we don't increment the reference count, we aren't preventing GC.

File: uarray/_backend.py (hunk 2 above)
Comment (peterbell10) [reply uarray/_backend.py]: I think you're describing the same issue in different words? You want to remove the context from the cache when the backend is collected, but it's existence in the cache prevents it from being collected.

File: uarray/_backend.py (hunk 1 above)
Comment (hameerabbasi) [reply uarray/_backend.py]: We're still keeping around references to invalid objects in a `unordered_map` or similar.

File: uarray/_backend.py (hunk 2 above)
Comment (hameerabbasi) [reply uarray/_backend.py]: We could cache the three `__ua_*__` attributes and release the backend.

Comment (hameerabbasi) [reply uarray/_backend.py]: Maybe the solution is to put the context in the backend's `__dict__` and release all other references to it?

Comment (hameerabbasi) [reply uarray/_backend.py]: ie. simply have a `__ua_cache__[("skip")|("set", coerce, only)]` attribute on the backend?

File: uarray/_backend.py (hunk 1 above)
Comment (peterbell10) [reply uarray/_backend.py]: You don't ever access the backend directly, only indirectly through the context. So, this doesn't matter.

Comment (hameerabbasi) [reply uarray/_backend.py]: It's still leaked memory.

File: uarray/_backend.py (hunk 2 above)
Comment (peterbell10) [reply uarray/_backend.py]: > We could cache the three __ua_*__ attributes and release the backend.

I think this just moves the problem to the attributes.

> Maybe the solution is to put the context in the backend's __dict__ and release all other references to it?

Good idea, I think that solves both problems doesn't it.

Comment (peterbell10) [reply uarray/_backend.py]: i.e. the backend owning a circular reference to itself doesn't matter so we  don't need `weakref` any more.

Comment (hameerabbasi) [reply uarray/_backend.py]: Opened #210 to that effect.

**Comment by hameerabbasi** [comment]
Closing in favour of #210.

**Pull request closed without merging**